ddl/index.go

```diff
@@ -700,6 +685,439 @@ func (w *worker) doBackfillIndexTaskInTxn(t table.Table, txn kv.Transaction, col
+		result.scanCount += scanCount
+		if result.scanCount-lastLogCount >= 30000 {
+			lastLogCount = result.scanCount
+			log.Infof("[ddl-reorg] worker(%v), finish batch addedCount:%v backfill, task addedCount:%v, task scanCount:%v, nextHandle:%v",
+				w.id, addedCount, result.addedCount, result.scanCount, nextHandle)
+		}
+
+		handleRange.startHandle = nextHandle
+		if handleRange.startHandle >= handleRange.endHandle {
+			break
+		}
+	}
+
+	return result
+}
+
+func (w *addIndexWorker) run() {
+	for {
+		task, more := <-w.taskCh
+		if !more {
+			break
+		}
+
+		log.Debug("[ddl-reorg] got backfill index task:#v", task)
+		result := w.handleBackfillTask(task)
+		w.resultCh <- result
+	}
+}
+
+func makeupIndexColFieldMap(t table.Table, indexInfo *model.IndexInfo) map[int64]*types.FieldType {
+	cols := t.Cols()
+	colFieldMap := make(map[int64]*types.FieldType, len(indexInfo.Columns))
+	for _, v := range indexInfo.Columns {
+		col := cols[v.Offset]
+		colFieldMap[col.ID] = &col.FieldType
+	}
+	return colFieldMap
+}
+
+// splitTableRanges use pd region's key ranges to split the backfilling table key range space,
+// to speed up adding index in table with disperse handle.
+func (d *ddl) splitTableRanges(t table.Table, startHandle int64) ([]kv.KeyRange, error) {
+	startRecordKey := t.RecordKey(startHandle)
+	endRecordKey := t.RecordKey(math.MaxInt64)
+	kvRange := kv.KeyRange{StartKey: startRecordKey, EndKey: endRecordKey}
+	store, ok := d.store.(tikv.Storage)
+	if !ok {
+		// Only support split ranges in tikv.Storage now.
+		return []kv.KeyRange{kvRange}, nil
+	}
+
+	maxSleep := 10000 // ms
+	bo := tikv.NewBackoffer(context.Background(), maxSleep)
+	ranges, err := tikv.SplitRegionRanges(bo, store.GetRegionCache(), []kv.KeyRange{kvRange})
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	if len(ranges) == 0 {
+		return nil, errors.Trace(errInvalidSplitRegionRanges)
+	}
+	return ranges, nil
+}
+
+func decodeHandleRange(keyRange kv.KeyRange) (int64, int64, error) {
+	_, startHandle, err := tablecodec.DecodeRecordKey(keyRange.StartKey)
+	if err != nil {
+		return 0, 0, errors.Trace(err)
+	}
+	_, endHandle, err := tablecodec.DecodeRecordKey(keyRange.EndKey)
+	if err != nil {
+		return 0, 0, errors.Trace(err)
+	}
+
+	return startHandle, endHandle, nil
+}
+
+func closeAddIndexWorkers(workers []*addIndexWorker) {
+	for _, worker := range workers {
+		worker.close()
+	}
+}
+
+func (d *ddl) waitTaskResults(workers []*addIndexWorker, taskCnt int, totalAddedCount *int64, startHandle int64) (int64, int64, error) {
+	var (
+		addedCount int64
+		nextHandle = startHandle
+		firstErr   error
+	)
+	for i := 0; i < taskCnt; i++ {
+		worker := workers[i]
+		result := <-worker.resultCh
+		if firstErr == nil && result.err != nil {
+			firstErr = result.err
+			log.Infof("[ddl-reorg] error occured, notify remaining workers to return.")
+
+			// Notify remaining workers exit immediately.
+			// If we don't do this, we will need to wait the long run workers exit.
```

Comment:
Why we need to the worker exit immediately? I think we exit worker takes a long time is OK.
And These codes make the logic more complicated.

Reply:
The worker checks if the job is runnable, if it's canceled by admin, all workers will exit immediately without the need to notify them.

Reply:
A shared variable and the runnable channel have the same effect.